xla/backends/gpu/runtime/all_to_all_thunk.cc

```diff
@@ -124,20 +131,75 @@ absl::Status AllToAllStartThunk::Initialize(const InitializeParams& params) {
     se::StreamExecutor* executor = params.executor;
     {
       absl::MutexLock lock(&pointer_maps_mutex_);
-      if (!send_pointer_maps_.count(executor)) {
+      if (!receive_pointer_maps_.count(executor)) {
         TF_ASSIGN_OR_RETURN(
             std::unique_ptr<se::MemoryAllocation> alloc,
             executor->HostMemoryAllocate(num_ranks * sizeof(uint64_t)));
         bool inserted =
-            send_pointer_maps_.insert({executor, std::move(alloc)}).second;
-        CHECK(inserted);
-        TF_ASSIGN_OR_RETURN(
-            alloc, executor->HostMemoryAllocate(num_ranks * sizeof(uint64_t)));
-        inserted =
             receive_pointer_maps_.insert({executor, std::move(alloc)}).second;
         CHECK(inserted);
       }
     }
+    {
+      absl::MutexLock lock(&events_mutex_);
+      if (!events_.count(executor)) {
+        TF_ASSIGN_OR_RETURN(std::unique_ptr<se::Event> event,
+                            executor->CreateEvent());
+        events_.insert({executor, std::move(event)});
+      }
+    }
+    std::optional<RankId> rank =
+        comm_handle.clique_key.rank(params.collective_params->global_device_id);
+    TF_ASSIGN_OR_RETURN(GpuCollectives * collectives,
+                        GetGpuCollectives(params));
+    size_t chunk_element_count = buffers_[0].element_count / num_ranks;
+    TF_ASSIGN_OR_RETURN(
+        std::vector<DeviceBufferPair> device_buffers,
+        ConvertToDeviceBuffers(params.buffer_allocations, buffers_,
+                               config_.config.operand_element_type));
+    for (int peer = 0; peer < num_ranks; ++peer) {
+      BufferRendezvousValue buffer_rendezvous_value;
+      buffer_rendezvous_value.rank = rank.value().value();
+      int buffer_idx = (rank.value().value() + peer) % num_ranks;
+      if (config_.has_split_dimension) {
+        buffer_rendezvous_value.buffer = reinterpret_cast<uint64_t>(
+            collectives
+                ->Slice(
+                    se::DeviceMemoryBase(device_buffers[0].destination_buffer),
+                    device_buffers[0].element_type,
```

Comment:
You call `device_buffers[0]` but can there be multiple device_buffers in the split_dimension case? Looks like the old logic does an all-to-all for each device buffer in the split_dimension case.

Reply:
> Yes that's true. Do you know in which case we would want to have multiple device buffer for split-dim all-to-all? I was never able to trigger it with existing tests and can't think of a case for it. Plus the other mode doesn't seem to support this while they should be identical in terms of functionality. I'd prefer legalizing the buffer size if that's never used.

You're right, the split_dim case seems to require a single operand and output. The verifier checks this. So using device_buffers[0] is fine.